File: apps/backend/core/src/main/java/com/schemafy/core/erd/service/MemoService.java
package com.schemafy.core.erd.service;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;

import org.springframework.stereotype.Service;

import com.schemafy.core.common.security.principal.AuthenticatedUser;
import com.schemafy.core.erd.controller.dto.request.CreateMemoCommentRequest;
import com.schemafy.core.erd.controller.dto.request.CreateMemoRequest;
import com.schemafy.core.erd.controller.dto.request.UpdateMemoCommentRequest;
import com.schemafy.core.erd.controller.dto.request.UpdateMemoRequest;
import com.schemafy.core.erd.controller.dto.response.MemoCommentResponse;
import com.schemafy.core.erd.controller.dto.response.MemoDetailResponse;
import com.schemafy.core.erd.controller.dto.response.MemoResponse;
import com.schemafy.core.erd.service.memo.MemoApiCommandMapper;
import com.schemafy.core.erd.service.memo.MemoApiResponseMapper;
import com.schemafy.core.user.controller.dto.response.UserSummaryResponse;
import com.schemafy.core.user.repository.UserRepository;
import com.schemafy.core.user.repository.entity.User;
import com.schemafy.domain.erd.memo.application.port.in.CreateMemoCommentUseCase;
import com.schemafy.domain.erd.memo.application.port.in.CreateMemoUseCase;
import com.schemafy.domain.erd.memo.application.port.in.DeleteMemoCommentUseCase;
import com.schemafy.domain.erd.memo.application.port.in.DeleteMemoUseCase;
import com.schemafy.domain.erd.memo.application.port.in.GetMemoCommentsUseCase;
import com.schemafy.domain.erd.memo.application.port.in.GetMemoUseCase;
import com.schemafy.domain.erd.memo.application.port.in.GetMemosBySchemaIdUseCase;
import com.schemafy.domain.erd.memo.application.port.in.UpdateMemoCommentUseCase;
import com.schemafy.domain.erd.memo.application.port.in.UpdateMemoPositionUseCase;
import com.schemafy.domain.erd.memo.domain.Memo;
import com.schemafy.domain.erd.memo.domain.MemoComment;
import com.schemafy.domain.erd.memo.domain.MemoDetail;

import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class MemoService {

  private final CreateMemoUseCase createMemoUseCase;
  private final GetMemoUseCase getMemoUseCase;
  private final GetMemosBySchemaIdUseCase getMemosBySchemaIdUseCase;
  private final UpdateMemoPositionUseCase updateMemoPositionUseCase;
  private final DeleteMemoUseCase deleteMemoUseCase;
  private final CreateMemoCommentUseCase createMemoCommentUseCase;
  private final GetMemoCommentsUseCase getMemoCommentsUseCase;
  private final UpdateMemoCommentUseCase updateMemoCommentUseCase;
  private final DeleteMemoCommentUseCase deleteMemoCommentUseCase;
  private final UserRepository userRepository;
  private final MemoApiCommandMapper commandMapper;
  private final MemoApiResponseMapper responseMapper;

  public Mono<MemoDetailResponse> createMemo(CreateMemoRequest request,
      AuthenticatedUser user) {
    return createMemoUseCase.createMemo(
        commandMapper.toCreateMemoCommand(request, user))
        .flatMap(this::buildMemoDetailResponse);
  }

  public Mono<MemoDetailResponse> getMemo(String memoId) {
    return getMemoUseCase.getMemo(
        commandMapper.toGetMemoQuery(memoId))
        .flatMap(this::buildMemoDetailResponse);
  }

  public Flux<MemoResponse> getMemosBySchemaId(String schemaId) {
    return getMemosBySchemaIdUseCase.getMemosBySchemaId(
        commandMapper.toGetMemosBySchemaIdQuery(schemaId))
        .collectList()
        .flatMapMany(this::mapMemosToResponses);
  }

  public Mono<MemoResponse> updateMemo(String memoId,
      UpdateMemoRequest request, AuthenticatedUser user) {
    return updateMemoPositionUseCase.updateMemoPosition(
        commandMapper.toUpdateMemoPositionCommand(memoId, request, user))
        .flatMap(memo -> getUserSummary(memo.authorId())
            .map(author -> responseMapper.toMemoResponse(memo, author)));
  }

  public Mono<Void> deleteMemo(String memoId, AuthenticatedUser user) {
    return deleteMemoUseCase.deleteMemo(
        commandMapper.toDeleteMemoCommand(memoId, user));
  }

  public Mono<MemoCommentResponse> createComment(String memoId,
      CreateMemoCommentRequest request, AuthenticatedUser user) {
    return createMemoCommentUseCase.createMemoComment(
        commandMapper.toCreateMemoCommentCommand(memoId, request, user))
        .flatMap(comment -> getUserSummary(comment.authorId())
            .map(author -> responseMapper.toMemoCommentResponse(comment,
                author)));
  }

  public Flux<MemoCommentResponse> getComments(String memoId) {
    return getMemoCommentsUseCase.getMemoComments(
        commandMapper.toGetMemoCommentsQuery(memoId))
        .collectList()
        .flatMapMany(this::mapCommentsToResponses);
  }

  public Mono<MemoCommentResponse> updateComment(String memoId,
      String commentId, UpdateMemoCommentRequest request,
      AuthenticatedUser user) {
    return updateMemoCommentUseCase.updateMemoComment(
        commandMapper.toUpdateMemoCommentCommand(memoId, commentId, request,
            user))
        .flatMap(comment -> getUserSummary(comment.authorId())
            .map(author -> responseMapper.toMemoCommentResponse(comment,
                author)));
  }

  public Mono<Void> deleteComment(String memoId, String commentId,
      AuthenticatedUser user) {
    return deleteMemoCommentUseCase.deleteMemoComment(
        commandMapper.toDeleteMemoCommentCommand(memoId, commentId, user));
  }

  private Mono<MemoDetailResponse> buildMemoDetailResponse(MemoDetail detail) {
    Set<String> authorIds = detail.comments().stream()
        .map(MemoComment::authorId)
        .collect(Collectors.toSet());
    authorIds.add(detail.memo().authorId());

    return loadUserSummaryMap(authorIds)
        .map(userMap -> {
          List<MemoCommentResponse> comments = detail.comments().stream()
              .map(comment -> responseMapper.toMemoCommentResponse(comment,
                  getUserFromMap(userMap, comment.authorId())))
              .toList();

          return MemoDetailResponse.builder()
              .id(detail.memo().id())
              .schemaId(detail.memo().schemaId())
              .author(getUserFromMap(userMap, detail.memo().authorId()))
              .positions(detail.memo().positions())
              .createdAt(detail.memo().createdAt())
              .updatedAt(detail.memo().updatedAt())
              .comments(comments)
              .build();
        });
  }

  private Flux<MemoResponse> mapMemosToResponses(List<Memo> memos) {
    if (memos.isEmpty()) {
      return Flux.empty();
    }

    Set<String> authorIds = memos.stream()
        .map(Memo::authorId)
        .collect(Collectors.toSet());

    return loadUserSummaryMap(authorIds)
        .flatMapMany(userMap -> Flux.fromIterable(memos)
            .map(memo -> responseMapper.toMemoResponse(memo,
                getUserFromMap(userMap, memo.authorId()))));
  }

  private Flux<MemoCommentResponse> mapCommentsToResponses(
      List<MemoComment> comments) {
    if (comments.isEmpty()) {
      return Flux.empty();
    }

    Set<String> authorIds = comments.stream()
        .map(MemoComment::authorId)
        .collect(Collectors.toSet());

    return loadUserSummaryMap(authorIds)
        .flatMapMany(userMap -> Flux.fromIterable(comments)
            .map(comment -> responseMapper.toMemoCommentResponse(comment,
                getUserFromMap(userMap, comment.authorId()))));
  }

  private Mono<Map<String, UserSummaryResponse>> loadUserSummaryMap(
      Set<String> userIds) {
    if (userIds.isEmpty()) {
      return Mono.just(Collections.emptyMap());
    }
    return userRepository.findAllById(userIds)
        .collectMap(User::getId, UserSummaryResponse::from);
  }

  private Mono<UserSummaryResponse> getUserSummary(String userId) {
    return userRepository.findById(userId)
        .map(UserSummaryResponse::from)
        .defaultIfEmpty(new UserSummaryResponse(userId, "Unknown"));
  }

  private UserSummaryResponse getUserFromMap(
      Map<String, UserSummaryResponse> userMap, String userId) {
    return userMap.getOrDefault(userId,
        new UserSummaryResponse(userId, "Unknown"));
  }

}

File: apps/bff/src/memo/memo.service.ts
import { Injectable } from '@nestjs/common';
import { BackendClientService } from '../common/backend-client/backend-client.service';
import {
  CreateMemoCommentRequest,
  CreateMemoRequest,
  Memo,
  MemoComment,
  UpdateMemoCommentRequest,
  UpdateMemoRequest,
} from './memo.types';

@Injectable()
export class MemoService {
  constructor(private readonly backendClient: BackendClientService) {}

  async createMemo(data: CreateMemoRequest, authHeader: string): Promise<Memo> {
    const response = await this.backendClient.client.post<Memo>(
      '/api/v1.0/memos',
      data,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async getMemo(memoId: string, authHeader: string): Promise<Memo> {
    const response = await this.backendClient.client.get<Memo>(
      `/api/v1.0/memos/${memoId}`,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async getSchemaMemos(schemaId: string, authHeader: string): Promise<Memo[]> {
    const response = await this.backendClient.client.get<Memo[]>(
      `/api/v1.0/schemas/${schemaId}/memos`,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async getSchemaMemosWithComments(
    schemaId: string,
    authHeader: string,
  ): Promise<Memo[]> {
    const memosResponse = await this.backendClient.client.get<Memo[]>(
      `/api/v1.0/schemas/${schemaId}/memos`,
      this.backendClient.getAuthConfig(authHeader),
    );

    const memos = memosResponse.data;

    const commentsResults = await Promise.allSettled(
      memos.map((memo) =>
        this.backendClient.client.get<MemoComment[]>(
          `/api/v1.0/memos/${memo.id}/comments`,
          this.backendClient.getAuthConfig(authHeader),
        ),
      ),
    );

    const memosWithComments = memos.map((memo, index) => {
      const result = commentsResults[index];
      const comments = result.status === 'fulfilled' ? result.value.data : [];
      return { ...memo, comments };
    });

    return memosWithComments;
  }

  async updateMemo(
    memoId: string,
    data: UpdateMemoRequest,
    authHeader: string,
  ): Promise<Memo> {
    const response = await this.backendClient.client.put<Memo>(
      `/api/v1.0/memos/${memoId}`,
      data,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async deleteMemo(memoId: string, authHeader: string): Promise<null> {
    const response = await this.backendClient.client.delete<null>(
      `/api/v1.0/memos/${memoId}`,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async createMemoComment(
    memoId: string,
    data: CreateMemoCommentRequest,
    authHeader: string,
  ): Promise<MemoComment> {
    const response = await this.backendClient.client.post<MemoComment>(
      `/api/v1.0/memos/${memoId}/comments`,
      data,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async getMemoComments(
    memoId: string,
    authHeader: string,
  ): Promise<MemoComment[]> {
    const response = await this.backendClient.client.get<MemoComment[]>(
      `/api/v1.0/memos/${memoId}/comments`,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async updateMemoComment(
    memoId: string,
    commentId: string,
    data: UpdateMemoCommentRequest,
    authHeader: string,
  ): Promise<MemoComment> {
    const response = await this.backendClient.client.put<MemoComment>(
      `/api/v1.0/memos/${memoId}/comments/${commentId}`,
      data,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }

  async deleteMemoComment(
    memoId: string,
    commentId: string,
    authHeader: string,
  ): Promise<null> {
    const response = await this.backendClient.client.delete<null>(
      `/api/v1.0/memos/${memoId}/comments/${commentId}`,
      this.backendClient.getAuthConfig(authHeader),
    );
    return response.data;
  }
}
